src/rgw/rgw_op.cc

```diff
@@ -2389,16 +2389,39 @@ void RGWGetObj::execute(optional_yield y)
     if (get_type() == RGW_OP_GET_OBJ && get_data && m.get_tier_type() == "cloud-s3") {
       RGWObjTier tier_config;
       m.get_tier_config(&tier_config);
-      op_ret = handle_cloudtier_obj(s, this, driver, attrs, sync_cloudtiered, tier_config, y);
-      if (op_ret < 0) {
-        ldpp_dout(this, 4) << "Cannot get cloud tiered object: " << *s->object
+      if (tier_config.tier_placement.allow_read_through) {
+        uint64_t days = tier_config.tier_placement.read_through_restore_days;
+        op_ret = handle_cloudtier_obj(s, this, driver, days, attrs, sync_cloudtiered, tier_config, y);
```

Comment:
I think restore_status should be checked before calling this routine. 

case1) if restore_status not set, restore the object.
case 2) if restore_status is RestoreAlreadyInProgress, return ERR_INVALID_OBJECT_STATE error. 
case 3) But if restore_status is CloudRestored, do not again restore the object but just return data.

In case (2) return error only if its GET request. If its HEAD request, do not return error but send attrs as response.